components/rc_log/src/android.rs

```diff
@@ -0,0 +1,209 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! This is th android backend for rc_log. It has a decent amount of
+//! complexity, as Rust logs can be emitted by any thread, regardless of whether
+//! or not they have an associated JVM thread. JNA's Callback class helps us
+//! here, by providing a way for mapping native threads to JVM threads.
+//! Unfortunately, naive usage of this class in a multithreaded context will be
+//! very suboptimal in terms of memory and thread usage.
+//!
+//! To avoid this, we only call into the JVM from a single thread, which we
+//! launch when initializing the logger. This thread just polls a channel
+//! listening for log messages, where a log message is an enum (`LogMessage`)
+//! that either tells it to log an item, or to stop logging all together.
+//!
+//! 1. We cannot guarantee the the callback from android lives past when the
+//!    android code tells us to stop logging, so in order to be memory safe, we
+//!    need to stop logging immediately when this happens. We do this using an
+//!    `Arc<AtomicBool>`, used to indicate that we should stop logging.
+//!
+//! 2. There's no safe way to terminate a thread in Rust (for good reason), so
+//!    the background thread must close willingly. To make sure this happens
+//!    promptly (e.g. to avoid a case where we're blocked until some thread
+//!    somewhere else happens to log something), we need to add something onto
+//!    the log channel, hence the existence of `LogMessage::Stop`.
+//!
+//!    It's important to note that because of point 1, the polling thread may
+//!    have to stop prior to getting `LogMessage::Stop`. We do not want to wait
+//!    for it to process whatever log messages were sent prior to being told to
+//!    stop.
+
+use std::{
+    ffi::CString,
+    os::raw::c_char,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        mpsc::{sync_channel, SyncSender},
+        Arc, Mutex,
+    },
+    thread,
+};
+
+use crate::LogLevel;
+
+/// Type of the log callback provided to us by java/swift. Takes the following
+/// arguments:
+///
+/// - Log level (an i32).
+/// - Tag: a (nullable) nul terminated c string. Caller must not free this
+///   string!
+/// - Message: a (non-nullable) nul terminated c string. Caller must not free
+///   this string!
+///
+/// and returns 0 if we should close the thread, and 1 otherwise. This is done
+/// because attempting to call `disable` from within the log callback will
+/// deadlock.
```

Comment:
It's so that we can't be dropped while the callback is running. Consider:

1. We emit a log, and are about to call the callback.
2. Before we do, some thread turns off the logger, and a garbage collection runs
3. We call the callback, which is now stale, boom.

Or something along those lines.

It's worth noting that all this android code is the same as before, just moved to a new file.